src/Core/Grid/Query/OrderQueryBuilder.php

```diff
@@ -0,0 +1,275 @@
+            ->addSelect('o.id_order, o.reference, o.total_paid_tax_incl, os.paid, osl.name AS osname')
+            ->addSelect('CONCAT(LEFT(cu.`firstname`, 1), \'. \', cu.`lastname`) AS `customer`')
+            ->addSelect('os.color, o.payment')
+            ->addSelect('o.date_add, cu.company, cl.name AS country_name, o.invoice_number, o.delivery_number')
+            ->addSelect('IF ((' . $newCustomerSubSelect->getSQL() . ') > 0, 0, 1) AS new')
+        ;
+
+        $qb = $this->applyNewCustomerFilter($qb, $searchCriteria->getFilters());
+
+        $this->criteriaApplicator
+            ->applyPagination($searchCriteria, $qb)
+        ;
+
+        $this->applySorting($qb, $searchCriteria);
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        return $this->connection
+            ->createQueryBuilder()
+            ->select('FOUND_ROWS()')
+        ;
+    }
+
+    /**
+     * @param array $filters
+     *
+     * @return QueryBuilder
+     */
+    private function getBaseQueryBuilder(array $filters)
+    {
+        $qb = $this->connection
+            ->createQueryBuilder()
+            ->from($this->dbPrefix . 'orders', 'o')
+            ->leftJoin('o', $this->dbPrefix . 'customer', 'cu', 'o.id_customer = cu.id_customer')
+            ->innerJoin('o', $this->dbPrefix . 'address', 'a', 'o.id_address_delivery = a.id_address')
+            ->innerJoin('a', $this->dbPrefix . 'country', 'c', 'a.id_country = c.id_country')
+            ->innerJoin(
+                'c',
+                $this->dbPrefix . 'country_lang',
+                'cl',
+                'c.id_country = cl.id_country AND cl.id_lang = :context_lang_id'
+            )
+            ->leftJoin('o', $this->dbPrefix . 'order_state', 'os', 'o.current_state = os.id_order_state')
+            ->leftJoin(
+                'os',
+                $this->dbPrefix . 'order_state_lang',
+                'osl',
+                'os.id_order_state = osl.id_order_state AND osl.id_lang = :context_lang_id'
+            )
+            ->leftJoin('o', $this->dbPrefix . 'shop', 's', 'o.id_shop = s.id_shop')
+            ->setParameter('context_lang_id', $this->contextLangId, PDO::PARAM_INT)
+        ;
+
+        $strictComparisonFilters = [
+            'id_order' => 'o.id_order',
+            'country_name' => 'c.id_country',
+            'total_paid_tax_incl' => 'o.total_paid_tax_incl',
+            'osname' => 'os.id_order_state',
+        ];
+
+        $likeComparisonFilters = [
+            'reference' => 'o.`reference`',
+            'company' => 'cu.`company`',
+            'customer' => 'cu.`lastname`',
+            'payment' => 'o.`payment`',
+        ];
+
+        $havingLikeComparisonFilters = [
+            'customer' => 'customer',
+        ];
+
+        $dateComparisonFilters = [
+            'date_add' => 'o.`date_add`',
+        ];
```

Comment:
@matks writing queries manually is repetitive along all grids. In addition, configuring filters for each grid is very painful and debugging to make sure they work well is time consuming whenever we do it.

It's also a copy/paste approach, since most of the time we copy code from another query builder and apply some changes. 

I think this could be a first step towards easier filters configuration. Wdyt?

It could be improved by extracting below `foreach` into separate service like:

```php
$filtersApplicator = new FiltersApplicator();

$config = [
    'date_filters' => [ ... ],
    'strict_filters' => [ ... ],
    ...
];

$filtersApplicator->apply($queryBuilder, $config);
```

Of course it's not well thought at this point, but this is something we could improve thus easing grids migration / creation.

@mickaelandrieu I dare to ping you on this as well. :stuck_out_tongue:

Reply:
I think it's a good idea 👍. To make it easy and efficient to use I suggest we do the following:
- `FiltersApplicator` must provide an easy and simple API such as `addFilter()`, `addDateFilter()` ... it must aim for a Developer Experience similar to Symfony FormTypes
- we'll make a dedicated integration testsuite for it (probably with phpunit, behat will bring no value here) to make sure it is battle-tested because if we have a bug in it, it'll make QueryBuilders very hard to debug

Reply:
Ok, but this should be done in separate PR I think. :)

Reply:
Agreed, and I see you started it 😄 
We can merge this PR then refactor it 😉